security/pkg/credentialfetcher/plugin/gce_test.go

```diff
@@ -0,0 +1,319 @@
+	t.Helper()
+	for _, err := range errs {
+		if err == nil && expectedErr != nil || err != nil && expectedErr == nil {
+			t.Errorf("%s, GetPlatformCredential() returns err: %v, want: %v", id, err, expectedErr)
+		} else if err != nil && expectedErr != nil && err.Error() != expectedErr.Error() {
+			t.Errorf("%s, GetPlatformCredential() returns err: %v, want: %v", id, err, expectedErr)
+		}
+	}
+}
+
+func verifyToken(t *testing.T, id, jwtPath string, tokens []string, expectedToken string) {
+	t.Helper()
+	for i, token := range tokens {
+		// if expectedToken is not set, mock metadata server returns an auto-generated fake token.
+		wantToken := fmt.Sprintf("%s%d", fakeTokenPrefix, i+1)
+		if expectedToken != "" {
+			wantToken = expectedToken
+		}
+		if token != wantToken {
+			t.Errorf("%s, #%d call to GetPlatformCredential() returns token: %s, want: %s",
+				id, i, token, wantToken)
+		}
+		if jwtPath != "" && i+1 == len(tokens) {
+			// If this is the last round, and JWT path is set, check JWT in the file.
+			jwtFile, err := getJWTFromFile(jwtPath)
+			if err != nil {
+				t.Fatalf("%s, failed to read token from file %s: %v", id, jwtPath, err)
+			}
+			if jwtFile != wantToken {
+				t.Errorf("%s, %s has token %s, want %s", id, jwtPath, jwtFile, wantToken)
+			}
+		}
+	}
+}
+
+func TestGCEPlugin(t *testing.T) {
+	testCases := map[string]struct {
+		jwt           string
+		jwtPath       string
+		expectedToken string
+		expectedCall  int
+		expectedErr   error
+	}{
+		"get VM credential": {
+			jwt:           thirdPartyJwt,
+			jwtPath:       fmt.Sprintf("/tmp/security-pkg-credentialfetcher-plugin-gcetest-%s", uuid.New().String()),
+			expectedToken: thirdPartyJwt,
+			expectedCall:  1,
+		},
+		"jwt path not set": {
+			jwt:          thirdPartyJwt,
+			expectedCall: 1,
+			expectedErr:  fmt.Errorf("jwtPath is unset"),
+		},
+		"fetch credential multiple times": {
+			expectedCall: 5,
+			jwtPath:      fmt.Sprintf("/tmp/security-pkg-credentialfetcher-plugin-gcetest-%s", uuid.New().String()),
+		},
+	}
+
+	SetTokenRotation(false)
+	ms, err := StartMetadataServer()
+	if err != nil {
+		t.Fatalf("StartMetadataServer() returns err: %v", err)
+	}
+	t.Cleanup(func() {
+		ms.Stop()
+		SetTokenRotation(true)
+	})
+
+	for id, tc := range testCases {
+		p := GCEPlugin{
+			tokenCache: tc.jwt,
+			jwtPath:    tc.jwtPath,
+		}
+		if err := creatJWTFile(tc.jwtPath); err != nil {
+			t.Fatalf("%s, creatJWTFile() returns err: %v", id, err)
+		}
+		ms.Reset()
+		ms.setToken(tc.jwt)
+
+		tokens, errs := getTokenFromServer(t, &p, tc.expectedCall)
+
+		verifyError(t, id, errs, tc.expectedErr)
+		if tc.expectedErr == nil {
+			if ms.NumGetTokenCall() != tc.expectedCall {
+				t.Errorf("%s, metadata server receives %d calls, want %d",
+					id, ms.NumGetTokenCall(), tc.expectedCall)
+			}
+			if tc.jwtPath != "" {
+				verifyToken(t, id, tc.jwtPath, tokens, tc.expectedToken)
+			}
+		}
+	}
+}
+
+func TestTokenRotationJob(t *testing.T) {
```

Comment:
This test is flaky:
```
=== RUN   TestTokenRotationJob
=== RUN   TestTokenRotationJob/expired_token_needs_rotation
=== RUN   TestTokenRotationJob/token_with_no_expiration_time_needs_rotation
=== RUN   TestTokenRotationJob/invalid_token_needs_rotation
    gce_test.go:310: invalid token needs rotation, /tmp/security-pkg-credentialfetcher-plugin-gcetest-3fa56879-88e9-4dad-9b5c-f9e35518788d has token , want invalid-token-section-1.invalid-token-section-2
--- FAIL: TestTokenRotationJob (6.01s)
    --- PASS: TestTokenRotationJob/expired_token_needs_rotation (2.00s)
    --- PASS: TestTokenRotationJob/token_with_no_expiration_time_needs_rotation (2.00s)
    --- FAIL: TestTokenRotationJob/invalid_token_needs_rotation (2.00s)
FAIL

ERROR: exit status 1
```

See https://github.com/istio/istio/wiki/Test-Flakes#reproducing-test-flakes

Reply:
Thanks, this is very helpful!

I have run stress test on `TestGCEPlugin`, that is very stable. No failures in 10 minutes. This `TestGCEPlugin` has verification of the token file, so I remove the same verification from `TestTokenRotationJob`. `TestTokenRotationJob` is flaky at token file verification (the test rewrites the file every 0.5s). We can add retry to wait until token file is ready to check, but that increases test time.

Now both tests are stable.
```
stress ./plugin.test -test.v -test.run TestGCEPlugin
5s: 1067 runs so far, 0 failures
10s: 2226 runs so far, 0 failures
15s: 3420 runs so far, 0 failures
20s: 4624 runs so far, 0 failures
25s: 5823 runs so far, 0 failures
30s: 7023 runs so far, 0 failures
35s: 8153 runs so far, 0 failures
40s: 9307 runs so far, 0 failures
45s: 10466 runs so far, 0 failures
50s: 11633 runs so far, 0 failures
55s: 12810 runs so far, 0 failures
1m0s: 13983 runs so far, 0 failures
1m5s: 15153 runs so far, 0 failures
1m10s: 16321 runs so far, 0 failures
1m15s: 17532 runs so far, 0 failures
1m20s: 18718 runs so far, 0 failures
1m25s: 19888 runs so far, 0 failures
1m30s: 21095 runs so far, 0 failures
1m35s: 22280 runs so far, 0 failures
1m40s: 23303 runs so far, 0 failures
1m45s: 24438 runs so far, 0 failures
1m50s: 25555 runs so far, 0 failures
1m55s: 26670 runs so far, 0 failures
2m0s: 27703 runs so far, 0 failures
2m5s: 28822 runs so far, 0 failures
2m10s: 29892 runs so far, 0 failures
2m15s: 30970 runs so far, 0 failures
2m20s: 32101 runs so far, 0 failures
2m25s: 33274 runs so far, 0 failures
2m30s: 34299 runs so far, 0 failures
2m35s: 35505 runs so far, 0 failures
2m40s: 36631 runs so far, 0 failures
2m45s: 37588 runs so far, 0 failures
2m50s: 38575 runs so far, 0 failures
2m55s: 39586 runs so far, 0 failures
3m0s: 40744 runs so far, 0 failures
3m5s: 41888 runs so far, 0 failures
3m10s: 43061 runs so far, 0 failures
3m15s: 44057 runs so far, 0 failures
3m20s: 45021 runs so far, 0 failures
3m25s: 46015 runs so far, 0 failures
3m30s: 47085 runs so far, 0 failures
3m35s: 48279 runs so far, 0 failures
3m40s: 49457 runs so far, 0 failures
3m45s: 50635 runs so far, 0 failures
3m50s: 51724 runs so far, 0 failures
3m55s: 52859 runs so far, 0 failures
4m0s: 54004 runs so far, 0 failures
4m5s: 55202 runs so far, 0 failures
4m10s: 56380 runs so far, 0 failures
4m15s: 57533 runs so far, 0 failures
4m20s: 58530 runs so far, 0 failures
4m25s: 59695 runs so far, 0 failures
4m30s: 60787 runs so far, 0 failures
4m35s: 61997 runs so far, 0 failures
4m40s: 63188 runs so far, 0 failures
4m45s: 64362 runs so far, 0 failures
4m50s: 65471 runs so far, 0 failures
4m55s: 66629 runs so far, 0 failures
5m0s: 67793 runs so far, 0 failures
5m5s: 68941 runs so far, 0 failures
5m10s: 70043 runs so far, 0 failures
5m15s: 71208 runs so far, 0 failures
5m20s: 72348 runs so far, 0 failures
5m25s: 73509 runs so far, 0 failures
5m30s: 74507 runs so far, 0 failures
5m35s: 75643 runs so far, 0 failures
5m40s: 76780 runs so far, 0 failures
5m45s: 77868 runs so far, 0 failures
5m50s: 78961 runs so far, 0 failures
5m55s: 80064 runs so far, 0 failures
6m0s: 81228 runs so far, 0 failures
6m5s: 82377 runs so far, 0 failures
6m10s: 83427 runs so far, 0 failures
6m15s: 84602 runs so far, 0 failures
6m20s: 85782 runs so far, 0 failures
6m25s: 86954 runs so far, 0 failures
6m30s: 88144 runs so far, 0 failures
6m35s: 89325 runs so far, 0 failures
6m40s: 90497 runs so far, 0 failures
6m45s: 91644 runs so far, 0 failures
6m50s: 92808 runs so far, 0 failures
6m55s: 93808 runs so far, 0 failures
7m0s: 94996 runs so far, 0 failures
7m5s: 96172 runs so far, 0 failures
7m10s: 97341 runs so far, 0 failures
7m15s: 98459 runs so far, 0 failures
7m20s: 99649 runs so far, 0 failures
7m25s: 100820 runs so far, 0 failures
7m30s: 102013 runs so far, 0 failures
7m35s: 103210 runs so far, 0 failures
7m40s: 104385 runs so far, 0 failures
7m45s: 105597 runs so far, 0 failures
7m50s: 106710 runs so far, 0 failures
7m55s: 107854 runs so far, 0 failures
8m0s: 109020 runs so far, 0 failures
8m5s: 110175 runs so far, 0 failures
8m10s: 111341 runs so far, 0 failures
8m15s: 112531 runs so far, 0 failures
8m20s: 113599 runs so far, 0 failures
8m25s: 114765 runs so far, 0 failures
8m30s: 115736 runs so far, 0 failures
8m35s: 116830 runs so far, 0 failures
8m40s: 118004 runs so far, 0 failures
8m45s: 119177 runs so far, 0 failures
8m50s: 120373 runs so far, 0 failures
8m55s: 121447 runs so far, 0 failures
9m0s: 122534 runs so far, 0 failures
9m5s: 123693 runs so far, 0 failures
9m10s: 124840 runs so far, 0 failures
9m15s: 126013 runs so far, 0 failures
9m20s: 127188 runs so far, 0 failures
9m25s: 128253 runs so far, 0 failures
9m30s: 129395 runs so far, 0 failures
9m35s: 130530 runs so far, 0 failures
9m40s: 131696 runs so far, 0 failures
9m45s: 132771 runs so far, 0 failures
9m50s: 133926 runs so far, 0 failures
9m55s: 135088 runs so far, 0 failures
10m0s: 136220 runs so far, 0 failures
10m5s: 137249 runs so far, 0 failures
10m10s: 138411 runs so far, 0 failures
```

```
stress ./plugin.test -test.v -test.run TestTokenRotationJob
5s: 0 runs so far, 0 failures
10s: 16 runs so far, 0 failures
15s: 32 runs so far, 0 failures
20s: 48 runs so far, 0 failures
25s: 64 runs so far, 0 failures
30s: 64 runs so far, 0 failures
35s: 80 runs so far, 0 failures
40s: 96 runs so far, 0 failures
45s: 112 runs so far, 0 failures
50s: 128 runs so far, 0 failures
55s: 144 runs so far, 0 failures
1m0s: 146 runs so far, 0 failures
1m5s: 160 runs so far, 0 failures
1m10s: 176 runs so far, 0 failures
1m15s: 192 runs so far, 0 failures
1m20s: 208 runs so far, 0 failures
1m25s: 224 runs so far, 0 failures
1m30s: 226 runs so far, 0 failures
1m35s: 240 runs so far, 0 failures
1m40s: 256 runs so far, 0 failures
1m45s: 272 runs so far, 0 failures
1m50s: 288 runs so far, 0 failures
1m55s: 296 runs so far, 0 failures
2m0s: 304 runs so far, 0 failures
2m5s: 320 runs so far, 0 failures
2m10s: 336 runs so far, 0 failures
2m15s: 352 runs so far, 0 failures
2m20s: 368 runs so far, 0 failures
2m25s: 370 runs so far, 0 failures
2m30s: 384 runs so far, 0 failures
2m35s: 400 runs so far, 0 failures
2m40s: 416 runs so far, 0 failures
2m45s: 432 runs so far, 0 failures
2m50s: 448 runs so far, 0 failures
2m55s: 451 runs so far, 0 failures
3m0s: 464 runs so far, 0 failures
3m5s: 480 runs so far, 0 failures
3m10s: 496 runs so far, 0 failures
3m15s: 512 runs so far, 0 failures
3m20s: 528 runs so far, 0 failures
3m25s: 531 runs so far, 0 failures
3m30s: 544 runs so far, 0 failures
3m35s: 560 runs so far, 0 failures
3m40s: 576 runs so far, 0 failures
3m45s: 592 runs so far, 0 failures
3m50s: 607 runs so far, 0 failures
3m55s: 612 runs so far, 0 failures
4m0s: 624 runs so far, 0 failures
4m5s: 640 runs so far, 0 failures
4m10s: 656 runs so far, 0 failures
4m15s: 672 runs so far, 0 failures
4m20s: 678 runs so far, 0 failures
4m25s: 689 runs so far, 0 failures
4m30s: 704 runs so far, 0 failures
4m35s: 720 runs so far, 0 failures
4m40s: 736 runs so far, 0 failures
4m45s: 752 runs so far, 0 failures
4m50s: 759 runs so far, 0 failures
4m55s: 768 runs so far, 0 failures
5m0s: 784 runs so far, 0 failures
5m5s: 800 runs so far, 0 failures
5m10s: 816 runs so far, 0 failures
5m15s: 832 runs so far, 0 failures
5m20s: 839 runs so far, 0 failures
5m25s: 848 runs so far, 0 failures
5m30s: 864 runs so far, 0 failures
5m35s: 880 runs so far, 0 failures
5m40s: 896 runs so far, 0 failures
5m45s: 912 runs so far, 0 failures
5m50s: 920 runs so far, 0 failures
5m55s: 928 runs so far, 0 failures
6m0s: 944 runs so far, 0 failures
6m5s: 960 runs so far, 0 failures
6m10s: 976 runs so far, 0 failures
6m15s: 990 runs so far, 0 failures
6m20s: 999 runs so far, 0 failures
6m25s: 1008 runs so far, 0 failures
6m30s: 1024 runs so far, 0 failures
6m35s: 1040 runs so far, 0 failures
6m40s: 1056 runs so far, 0 failures
6m45s: 1064 runs so far, 0 failures
6m50s: 1072 runs so far, 0 failures
6m55s: 1088 runs so far, 0 failures
7m0s: 1104 runs so far, 0 failures
7m5s: 1120 runs so far, 0 failures
7m10s: 1136 runs so far, 0 failures
7m15s: 1146 runs so far, 0 failures
7m20s: 1152 runs so far, 0 failures
7m25s: 1168 runs so far, 0 failures
7m30s: 1184 runs so far, 0 failures
7m35s: 1200 runs so far, 0 failures
7m40s: 1216 runs so far, 0 failures
7m45s: 1226 runs so far, 0 failures
7m50s: 1232 runs so far, 0 failures
7m55s: 1248 runs so far, 0 failures
8m0s: 1264 runs so far, 0 failures
8m5s: 1280 runs so far, 0 failures
8m10s: 1296 runs so far, 0 failures
8m15s: 1306 runs so far, 0 failures
8m20s: 1313 runs so far, 0 failures
8m25s: 1328 runs so far, 0 failures
8m30s: 1344 runs so far, 0 failures
8m35s: 1360 runs so far, 0 failures
8m40s: 1373 runs so far, 0 failures
8m45s: 1381 runs so far, 0 failures
8m50s: 1392 runs so far, 0 failures
8m55s: 1408 runs so far, 0 failures
9m0s: 1424 runs so far, 0 failures
9m5s: 1440 runs so far, 0 failures
9m10s: 1450 runs so far, 0 failures
9m15s: 1457 runs so far, 0 failures
9m20s: 1472 runs so far, 0 failures
9m25s: 1488 runs so far, 0 failures
9m30s: 1504 runs so far, 0 failures
9m35s: 1520 runs so far, 0 failures
9m40s: 1530 runs so far, 0 failures
9m45s: 1537 runs so far, 0 failures
9m50s: 1552 runs so far, 0 failures
9m55s: 1568 runs so far, 0 failures
10m0s: 1584 runs so far, 0 failures
10m5s: 1600 runs so far, 0 failures
10m10s: 1610 runs so far, 0 failures
10m15s: 1617 runs so far, 0 failures
10m20s: 1632 runs so far, 0 failures
```